File: issues/5010/test_issue_5010_large.py
#!/usr/bin/env python3
"""
复现 Faiss Issue #5010（大规模测试）
根据 Issue 描述，问题在于：
- 当 prefetch_nthread > 0 时，所有线程都在等待单个读操作完成
- 期望：多个列表并行预取
- 实际：串行等待，导致性能下降
"""

import faiss
import numpy as np
import time
import os
import tempfile

def test_large_scale():
    """大规模测试 - 更接近真实场景"""
    
    # 更大的参数（Issue 中使用的是百万级别）
    d = 128           # 维度
    nb = 1_000_000    # 100万向量
    nq = 1000         # 1000个查询
    k = 100           # top-100
    nlist = 4096      # 4096个聚类中心（更真实）
    nprobe = 64       # 搜索64个聚类
    
    print("=" * 60)
    print("Faiss Issue #5010 大规模复现测试")
    print("=" * 60)
    print(f"配置: d={d}, nb={nb:,}, nq={nq}, k={k}, nlist={nlist}, nprobe={nprobe}")
    print(f"数据大小: ~{nb * d * 4 / 1024 / 1024:.1f} MB")
    print()
    
    # 生成随机数据
    print("生成数据...")
    np.random.seed(42)
    xb = np.random.random((nb, d)).astype('float32')
    xq = np.random.random((nq, d)).astype('float32')
    
    # 创建索引
    print("1. 构建 IndexIVFFlat 索引...")
    quantizer = faiss.IndexFlatL2(d)
    index = faiss.IndexIVFFlat(quantizer, d, nlist)
    
    # 训练
    print("   训练中...")
    index.train(xb[:100000])  # 用10万样本训练
    
    # 转换为磁盘倒排列表
    temp_dir = tempfile.mkdtemp()
    invlists_file = os.path.join(temp_dir, "large_invlists.dat")
    print(f"2. 转换为 OnDiskInvertedLists: {invlists_file}")
    
    ondisk_lists = faiss.OnDiskInvertedLists(nlist, index.code_size, invlists_file)
    index.replace_invlists(ondisk_lists)
    
    # 添加数据
    print("3. 添加数据...")
    batch_size = 100000
    for i in range(0, nb, batch_size):
        end = min(i + batch_size, nb)
        index.add(xb[i:end])
        print(f"   已添加 {end:,}/{nb:,} 向量")
    
    # 保存并重新加载
    index_file = os.path.join(temp_dir, "large_index.bin")
    print(f"4. 保存并重新加载索引: {index_file}")
    faiss.write_index(index, index_file)
    
    print("   重新加载...")
    reloaded_index = faiss.read_index(index_file)
    reloaded_index.nprobe = nprobe
    
    # 检查磁盘文件大小
    disk_size_mb = os.path.getsize(invlists_file) / 1024 / 1024
    print(f"   磁盘倒排列表大小: {disk_size_mb:.1f} MB")
    
    print()
    print("=" * 60)
    print("性能测试（大规模）")
    print("=" * 60)
    
    # 清理系统缓存（在 macOS 上）
    print("清理系统缓存...")
    os.system("sync")
    os.system("purge 2>/dev/null")
    time.sleep(2)
    
    results = []
    for prefetch_threads in [0, 4, 16, 32]:
        # 设置预取线程数
        if hasattr(reloaded_index.invlists, 'prefetch_nthread'):
            reloaded_index.invlists.prefetch_nthread = prefetch_threads
        
        print(f"\n测试 prefetch_threads={prefetch_threads}...")
        
        # 热身（少量查询）
        _ = reloaded_index.search(xq[:10], k)
        
        # 多次测试取平均
        times = []
        for run in range(3):
            start = time.time()
            distances, labels = reloaded_index.search(xq, k)
            elapsed_ms = (time.time() - start) * 1000
            times.append(elapsed_ms)
            print(f"  运行 {run+1}: {elapsed_ms:.2f} ms")
        
        avg_time = np.mean(times)
        std_time = np.std(times)
        results.append((prefetch_threads, avg_time, std_time))
        print(f"  平均: {avg_time:.2f} ± {std_time:.2f} ms")
    
    # 分析结果
    print()
    print("=" * 60)
    print("结果分析（大规模）")
    print("=" * 60)
    
    baseline_time = results[0][1]
    print(f"基准时间 (threads=0): {baseline_time:.2f} ms")
    print()
    
    issue_detected = False
    for threads, avg_time, std_time in results[1:]:
        speedup = baseline_time / avg_time
        if speedup < 1.0:
            slowdown_pct = (avg_time / baseline_time - 1) * 100
            print(f"🔴 threads={threads:2d}: 性能下降 {slowdown_pct:+.1f}% - Issue #5010!")
            issue_detected = True
        elif speedup < 1.05:  # 提升不到5%也可能有问题
            print(f"⚠️  threads={threads:2d}: 微小提升 {(speedup-1)*100:+.1f}% (预期更高)")
        else:
            print(f"✅ threads={threads:2d}: 性能提升 {(speedup-1)*100:+.1f}%")
    
    print()
    if issue_detected:
        print("🔴 Issue #5010 已复现：多线程预取导致性能下降")
        print("   根据 Issue 描述，这是因为所有线程在等待单个磁盘读操作")
    else:
        print("✅ 未检测到 Issue #5010 症状（可能已修复或环境不同）")
    
    # 清理
    print()
    print(f"清理临时文件: {temp_dir}")
    os.remove(invlists_file)
    os.remove(index_file)
    os.rmdir(temp_dir)
    
    return results

if __name__ == "__main__":
    results = test_large_scale()

File: issues/5010/test_issue_5010_large.cpp
#include <faiss/IndexIVFFlat.h>
#include <faiss/IndexFlat.h>
#include <faiss/index_io.h>
#include <faiss/invlists/OnDiskInvertedLists.h>
#include <chrono>
#include <iostream>
#include <vector>
#include <random>
#include <iomanip>

int main() {
    // 增加数据规模以匹配原始 Issue
    int d = 384;      // all-MiniLM-L6-v2 维度
    int nb = 200000;  // 20万向量（平衡测试时间和数据规模）
    int nq = 1000;    // 1000个查询
    int k = 10;
    int nlist = 1024; // 增加到1024
    int nprobe = 64;  // 增加到64
    
    std::cout << "=== Faiss Issue #5010 - Large Scale Test ===" << std::endl;
    std::cout << "Parameters:" << std::endl;
    std::cout << "  - Dimensions: " << d << std::endl;
    std::cout << "  - Vectors: " << nb << std::endl;
    std::cout << "  - Queries: " << nq << std::endl;
    std::cout << "  - nlist: " << nlist << std::endl;
    std::cout << "  - nprobe: " << nprobe << std::endl;
    std::cout << std::endl;
    
    std::cout << "Building test index (this may take a minute)..." << std::endl;
    
    // 生成随机数据
    std::vector<float> xb(nb * d), xq(nq * d);
    std::mt19937 rng(42);
    std::uniform_real_distribution<> dist(0, 1);
    for (auto& v : xb) v = dist(rng);
    for (auto& v : xq) v = dist(rng);
    
    // 创建索引
    faiss::IndexFlatL2 quantizer(d);
    faiss::IndexIVFFlat index(&quantizer, d, nlist);
    
    std::cout << "Training index..." << std::endl;
    index.train(nb, xb.data());
    
    // 转换为磁盘索引
    std::cout << "Converting to on-disk index..." << std::endl;
    faiss::OnDiskInvertedLists* invlists = 
        new faiss::OnDiskInvertedLists(nlist, index.code_size, "/tmp/test_invlists_large.dat");
    index.replace_invlists(invlists, true);
    
    std::cout << "Adding vectors..." << std::endl;
    index.add(nb, xb.data());
    
    // 保存并重新加载
    std::cout << "Saving index..." << std::endl;
    faiss::write_index(&index, "/tmp/test_index_large.bin");
    
    std::cout << "Reloading index..." << std::endl;
    auto* reloaded = dynamic_cast<faiss::IndexIVFFlat*>(
        faiss::read_index("/tmp/test_index_large.bin"));
    reloaded->nprobe = nprobe;
    
    // 获取 OnDiskInvertedLists 指针
    auto* ondisk = dynamic_cast<faiss::OnDiskInvertedLists*>(reloaded->invlists);
    if (!ondisk) {
        std::cerr << "Error: Not an OnDiskInvertedLists!" << std::endl;
        return 1;
    }
    
    // 测试不同预取配置
    std::vector<int> configs = {0, 1, 4, 8, 16, 32};
    std::vector<float> distances(nq * k);
    std::vector<faiss::idx_t> labels(nq * k);
    
    std::cout << "\n=== Performance Test (20万向量) ===" << std::endl;
    std::cout << "| Prefetch Threads | Latency (ms) | Change     |" << std::endl;
    std::cout << "|------------------|--------------|------------|" << std::endl;
    
    double baseline = 0;
    for (int nt : configs) {
        // 设置预取线程数
        ondisk->prefetch_nthread = nt;
        
        // 预热（3次）
        for (int i = 0; i < 3; i++) {
            reloaded->search(10, xq.data(), k, distances.data(), labels.data());
        }
        
        // 实际测试（3次取平均，减少测试时间）
        const int num_runs = 3;
        double total_ms = 0;
        
        for (int run = 0; run < num_runs; run++) {
            auto start = std::chrono::high_resolution_clock::now();
            reloaded->search(nq, xq.data(), k, distances.data(), labels.data());
            auto end = std::chrono::high_resolution_clock::now();
            
            double ms = std::chrono::duration<double, std::milli>(end - start).count();
            total_ms += ms;
        }
        
        double avg_ms = total_ms / num_runs;
        
        if (nt == 0) {
            baseline = avg_ms;
            std::cout << "| No prefetch      | " 
                     << std::fixed << std::setprecision(2) << std::setw(9) << avg_ms 
                     << "  | baseline   |" << std::endl;
        } else {
            double pct = ((avg_ms - baseline) / baseline) * 100;
            std::cout << "| " << std::setw(2) << nt << " threads       | " 
                     << std::fixed << std::setprecision(2) << std::setw(9) << avg_ms 
                     << "  | " << std::showpos << std::setw(6) << pct << std::noshowpos << "% |" 
                     << std::endl;
        }
    }
    
    std::cout << "\n=== Test Complete ===" << std::endl;
    
    delete reloaded;
    return 0;
}
